src/transcripts.rs

```diff
@@ -1,76 +1,199 @@
+{
+    transcript: Transcript,
+    bytes: Option<Zeroizing<Vec<u8>>>,
+    _phantom: PhantomData<P>,
+}
+
+impl<P> RangeProofTranscript<P>
 where
     P: Compressable + Precomputable,
     P::Compressed: FixedBytesRepr + IsIdentity,
 {
-    transcript.validate_and_append_point(b"H", h_base_compressed)?;
-    for item in g_base_compressed {
-        transcript.validate_and_append_point(b"G", item)?;
+    /// Initialize a transcript.
+    ///
+    /// The prover should include its `witness` here; the verifier should pass `None`.
+    #[allow(clippy::too_many_arguments)]
+    pub(crate) fn new<R: CryptoRngCore>(
+        label: &'static str,
+        h_base_compressed: &P::Compressed,
+        g_base_compressed: &[P::Compressed],
+        bit_length: usize,
+        extension_degree: usize,
+        aggregation_factor: usize,
+        statement: &RangeStatement<P>,
+        witness: Option<&RangeWitness>,
+        external_rng: &mut R,
+    ) -> Result<(Self, TranscriptRng), ProofError> {
+        // Initialize the transcript with parameters and statement
+        let mut transcript = Transcript::new(label.as_bytes());
+        transcript.domain_separator(b"Bulletproofs+", b"Range Proof");
+        transcript.validate_and_append_point(b"H", h_base_compressed)?;
+        for item in g_base_compressed {
+            transcript.validate_and_append_point(b"G", item)?;
+        }
+        transcript.append_u64(b"N", bit_length as u64);
+        transcript.append_u64(b"T", extension_degree as u64);
+        transcript.append_u64(b"M", aggregation_factor as u64);
+        for item in &statement.commitments_compressed {
+            transcript.append_point(b"Ci", item);
+        }
+        for item in &statement.minimum_value_promises {
+            if let Some(minimum_value) = item {
+                transcript.append_u64(b"vi - minimum_value", *minimum_value);
+            } else {
+                transcript.append_u64(b"vi - minimum_value", 0);
+            }
+        }
+
+        // Serialize the witness if provided
+        let bytes = if let Some(witness) = witness {
+            let size: usize = witness
+                .openings
+                .iter()
+                .map(|o| size_of::<u64>() + o.r.len() * size_of::<Scalar>())
+                .sum();
+            let mut witness_bytes = Zeroizing::new(Vec::<u8>::with_capacity(size));
+            for opening in &witness.openings {
+                witness_bytes.extend(opening.v.to_le_bytes());
+                for r in &opening.r {
+                    witness_bytes.extend(r.as_bytes());
+                }
+            }
+
+            Some(witness_bytes)
+        } else {
+            None
+        };
+
+        // Set up the RNG
+        let transcript_rng = Self::build_rng(&transcript, bytes.as_ref(), external_rng);
+
+        Ok((
+            Self {
+                transcript,
+                bytes,
+                _phantom: PhantomData,
+            },
+            transcript_rng,
```

Comment:
(optional) Could we not encapsulate the transcript_rng in the `Transcript`?

For scalar generation, we could either expose the RNG
```rust
    pub(crate) fn rng_mut(&mut self) -> &mut TranscriptRng {
        &mut self.transcript_rng
    }
```
or perhaps

```rust
pub(crate) fn generate_random_scalar(&mut self) -> Scalar {
   Scalar::random_not_zero(&mut transcript.rng)
}
```

Reply:
Ooh, I like the first idea, since it keeps everything more nicely contained. I was even hoping to have `RangeProofTranscript` hold a mutable reference to the external RNG so the prover and verifier wouldn't be able to misuse it, but the lifetimes got messy and I decided against it.

There should be some sanity tests for the resulting `TranscriptRng`, so I might add this suggestion in a follow-up PR.